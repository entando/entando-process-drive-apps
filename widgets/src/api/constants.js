export const DOMAINS = {
  PDA: process.env.REACT_APP_DOMAIN || '',
  APP_BUILDER: process.env.REACT_APP_APP_BUILDER_DOMAIN || '',
};

export const IS_MOCKED_API = process.env.REACT_APP_MOCKED_API === 'true';

export const MOCK_API_DELAY = 800;

export const LOCAL = process.env.REACT_APP_LOCAL === 'true';

export const METHODS = {
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT',
  DELETE: 'DELETE',
};

// Inject ur access_token if you are trying to reach the APP-BUILDER API locally
export const TOKENS = {
  APP_BUILDER: '0694821a5bdc2229aac820e8a773669a',
};
