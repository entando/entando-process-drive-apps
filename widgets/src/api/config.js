import { DOMAIN, IS_MOCKED_API, MOCK_API_DELAY } from 'api/constants';
import utils from 'utils';
import config from 'mocks/config';

const get = async widgetCode => {
  if (IS_MOCKED_API) {
    await utils.timeout(MOCK_API_DELAY);
    return config;
  }

  const url = `${DOMAIN}/configs/${widgetCode}`;
  const response = await fetch(url);

  return response.json();
};

export default {
  get,
};
