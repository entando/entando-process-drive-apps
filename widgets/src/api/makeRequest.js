import { IS_MOCKED_API, MOCK_API_DELAY, DOMAINS, METHODS } from 'api/constants';
import utils from 'utils';

const getParams = queryParams =>
  Object.keys(queryParams)
    .filter(k => queryParams[k] !== undefined && queryParams[k] !== '')
    .map(k => `${encodeURIComponent(k)}=${encodeURIComponent(queryParams[k])}`)
    .join('&');

export default async ({
  domain,
  uri,
  method,
  mockResponse = {},
  useAuthentication,
  body,
  headers = {},
  queryParams,
  forceMock,
}) => {
  if (IS_MOCKED_API || forceMock) {
    await utils.timeout(MOCK_API_DELAY);
    return mockResponse;
  }

  const url = `${domain || DOMAINS.PDA}${uri}${queryParams ? `?${getParams(queryParams)}` : ''}`;

  const requestHeaders = useAuthentication
    ? new Headers({
        'Content-Type': 'application/json',
        Authorization: `Bearer ${localStorage.getItem('token')}`,
        ...headers,
      })
    : new Headers(headers);

  const configs = {
    method: method || METHODS.GET,
    ...(body ? { body } : {}),
    headers: requestHeaders,
  };

  const response = await fetch(url, configs);
  if (!response.ok) {
    const payload = await response.json();
    throw new Error(
      payload.message
        ? payload.message
        : `Bad Response from server: ${response.status} ${response.statusText}`
    );
  }

  const responseHeaders = response.headers.get('Content-Type');

  if (responseHeaders && responseHeaders.includes('xml')) {
    return response.text();
  }
  if (responseHeaders && responseHeaders.includes('octet')) {
    return response;
  }
  return response.json();
};
