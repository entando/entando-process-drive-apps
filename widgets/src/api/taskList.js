import { IS_MOCKED_API, MOCK_API_DELAY, SERVICE } from 'api/constants';
import getTasks from 'mocks/taskList/api.mock';
import utils from 'utils';

const get = async (connection, page, pageSize, sortedColumn, sortOrder, filter) => {
  if (IS_MOCKED_API) {
    await utils.timeout(MOCK_API_DELAY);
    return getTasks(page, pageSize, sortedColumn, sortOrder, filter);
  }

  const url = `${SERVICE.URL}/connections/${connection}/tasks?sort=taskId`;
  const token = localStorage.getItem('token');
  const response = await fetch(url, {
    method: 'get',
    headers: new Headers({
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    }),
  });

  return response.json();
};

export default {
  get,
};
