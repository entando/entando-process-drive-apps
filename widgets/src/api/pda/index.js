import makeRequest from 'api/makeRequest';
import { SERVICE, METHODS } from 'api/constants';

import MOCK_CONNECTIONS from 'mocks/pda/connections';
import MOCK_PROCESSES from 'mocks/pda/processes';

export const getConnections = async () =>
  makeRequest({
    domain: SERVICE.URL,
    uri: '/connections',
    method: METHODS.GET,
    mockResponse: MOCK_CONNECTIONS,
    useAuthentication: false,
  });

export const getProcess = async connection =>
  makeRequest({
    domain: SERVICE.URL,
    uri: `/connections/${connection}/processes/definitions`,
    method: METHODS.GET,
    mockResponse: MOCK_PROCESSES,
    useAuthentication: false,
  });
