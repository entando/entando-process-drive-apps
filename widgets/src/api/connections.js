import { METHODS, DOMAINS } from 'api/constants';

import { CONNECTIONS } from 'mocks/taskList/configs';
import makeRequest from 'api/makeRequest';

// eslint-disable-next-line import/prefer-default-export
export const getConnections = async () =>
  makeRequest({
    domain: DOMAINS.PDA,
    uri: '/connections',
    method: METHODS.GET,
    mockResponse: CONNECTIONS,
    useAuthentication: false,
  });
