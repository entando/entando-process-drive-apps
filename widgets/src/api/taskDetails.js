import { SERVICE, METHODS } from 'api/constants';
import mockedTask from 'mocks/taskDetails/task';

import { CONNECTIONS, PROCESS, GROUPS, COLUMNS } from 'mocks/taskList/configs';
import makeRequest from 'api/makeRequest';

export const getTask = async (connection, taskId) => {
  return makeRequest({
    domain: SERVICE.URL,
    uri: `/connections/${connection}/tasks/${taskId}`,
    method: METHODS.GET,
    mockResponse: mockedTask,
    useAuthentication: true,
  });
};

// Configs

export const getConnections = async () =>
  makeRequest({
    domain: SERVICE.URL,
    uri: '/connections',
    method: METHODS.GET,
    mockResponse: CONNECTIONS,
    useAuthentication: false,
  });

export const getProcess = async connection =>
  makeRequest({
    // domain: PDA_DOMAIN,
    uri: `/connections/${connection}/processes/definitions`,
    method: METHODS.GET,
    mockResponse: PROCESS,
    useAuthentication: true,
  });

export const getGroups = async connection =>
  makeRequest({
    // domain: PDA_DOMAIN,
    uri: `/connections/${connection}/groups`,
    method: METHODS.GET,
    mockResponse: GROUPS,
    useAuthentication: true,
  });

export const getColumns = async connection =>
  makeRequest({
    // domain: PDA_DOMAIN,
    uri: `/connections/${connection}/tasks/columns`,
    method: METHODS.GET,
    mockResponse: COLUMNS,
    useAuthentication: true,
  });
