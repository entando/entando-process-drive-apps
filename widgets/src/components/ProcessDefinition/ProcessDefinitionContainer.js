import { ThemeProvider } from '@material-ui/core/styles';
import React from 'react';
import PropTypes from 'prop-types';
import Container from '@material-ui/core/Container';

import { getProcessForm, postProcessForm } from 'api/pda/processes';
import { getPageWidget } from 'api/app-builder/pages';
import theme from 'theme';
import CustomEventContext from 'components/common/CustomEventContext';
import WidgetBox from 'components/common/WidgetBox';
import JSONForm from 'components/common/form/JSONForm';
import Notification from 'components/common/Notification';
import withAuth from 'components/common/auth/withAuth';

function isJsonString(str) {
  try {
    JSON.parse(str);
  } catch (e) {
    return false;
  }
  return true;
}

class ProcessDefinitionContainer extends React.Component {
  state = {
    config: null,
    loading: false,
    submitting: false,
    formSchema: null,
    errorMessage: '',
  };

  componentDidMount() {
    this.setState({ loading: true }, async () => {
      const config = await this.fetchWidgetConfigs();
      if (config) {
        this.setState({ config, loading: false }, async () => {
          const formSchema = await this.fetchSchema();
          this.setState({ formSchema });
        });
      }
    });
  }

  fetchWidgetConfigs = async () => {
    const { pageCode, frameId } = this.props;
    try {
      // config will be fetched from app-builder
      const widgetConfigs = await getPageWidget(pageCode, frameId);
      if (widgetConfigs.errors && widgetConfigs.errors.length) {
        throw widgetConfigs.errors[0];
      }
      const { config } = widgetConfigs.payload;
      const settings = (config.settings && JSON.parse(config.settings)) || {};

      const parsedSettings = Object.keys(settings).reduce(
        (acc, settingKey) => ({
          ...acc,
          [settingKey]: isJsonString(settings[settingKey])
            ? JSON.parse(settings[settingKey])
            : settings[settingKey],
        }),
        {}
      );

      return {
        ...config,
        settings: parsedSettings,
      };
    } catch (error) {
      this.handleError(error.message);
    }
    return null;
  };

  closeNotification = () => {
    this.setState({ errorMessage: '' });
  };

  fetchSchema = async () => {
    const { config } = this.state;
    const { knowledgeSource = '', settings = {} } = config;
    const { processDefinition } = settings;

    try {
      const formSchema = await getProcessForm(knowledgeSource, processDefinition);
      return formSchema;
    } catch (error) {
      this.handleError(error.message);
    }
    return null;
  };

  submitProcessForm = form => {
    this.setState({ submitting: true }, async () => {
      const { config } = this.state;
      const { onSubmitForm } = this.props;
      const { knowledgeSource = '', settings = {} } = config;
      const { processDefinition } = settings;

      try {
        const response = await postProcessForm(knowledgeSource, processDefinition, form.formData);
        onSubmitForm({ ...form, response });
      } catch (error) {
        this.handleError(error.message);
      } finally {
        this.setState({ submitting: false });
      }
    });
  };

  handleError = errorMessage => {
    this.setState({ errorMessage });
    const { onError } = this.props;
    onError(errorMessage);
  };

  render() {
    const { loading, formSchema, config, submitting, errorMessage } = this.state;
    const { onError } = this.props;

    const uiSchemas = (config && config.settings && config.settings.uiSchemas) || [];

    return (
      <CustomEventContext.Provider
        value={{
          onSubmitForm: this.submitProcessForm,
          onError,
        }}
      >
        <ThemeProvider theme={theme}>
          <Container disableGutters>
            <WidgetBox>
              <JSONForm
                loading={loading}
                formSchema={formSchema}
                uiSchemas={uiSchemas}
                submitting={submitting}
              />
            </WidgetBox>
          </Container>
          <Notification message={errorMessage} type="error" onClose={this.closeNotification} />
        </ThemeProvider>
      </CustomEventContext.Provider>
    );
  }
}

ProcessDefinitionContainer.propTypes = {
  onError: PropTypes.func,
  onSubmitForm: PropTypes.func,
  pageCode: PropTypes.string,
  frameId: PropTypes.string,
};

ProcessDefinitionContainer.defaultProps = {
  onError: () => {},
  onSubmitForm: () => {},
  pageCode: '',
  frameId: '',
};

export default withAuth(ProcessDefinitionContainer, [
  'process-definition-form-get',
  'process-definition-form-submit',
]);
