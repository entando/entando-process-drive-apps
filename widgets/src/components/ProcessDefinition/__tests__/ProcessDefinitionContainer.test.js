import { render, wait } from '@testing-library/react';
import React from 'react';
import 'mocks/i18nMock';

import mockKeycloak from 'mocks/auth/keycloak';

import { DOMAINS } from 'api/constants';
import ProcessDefinitionContainer from 'components/ProcessDefinition/ProcessDefinitionContainer';
import WIDGET_CONFIGS from 'mocks/app-builder/pages';
import MOCK_PROCESSES from 'mocks/pda/processes';

mockKeycloak();

describe('<ProcessDefinitionContainer />', () => {
  it('renders snapshot correctly', async () => {
    const configUrl = `${DOMAINS.APP_BUILDER}/api/pages//widgets/`;
    const connection = 'kieStaging';
    const processUrl = `${DOMAINS.PDA}/connections/${connection}/processes/definitions`;

    fetch
      .once(JSON.stringify(WIDGET_CONFIGS.PROCESS_DEFINITION))
      .once(JSON.stringify(MOCK_PROCESSES));

    const { container } = render(<ProcessDefinitionContainer />);

    await wait(() => expect(fetch.mock.calls.length).toBe(2));

    expect(container).toMatchSnapshot();
    expect(fetch.mock.calls[0][0]).toEqual(configUrl);
    expect(fetch.mock.calls[1][0]).toEqual(processUrl);
  });

  it('renders snapshot correctly on error state', async () => {
    const { container } = render(<ProcessDefinitionContainer />);

    await wait(() => expect(container).toMatchSnapshot());
  });
});
