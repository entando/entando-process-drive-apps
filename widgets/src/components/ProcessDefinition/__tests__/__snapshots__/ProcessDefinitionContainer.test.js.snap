// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProcessDefinitionContainer /> renders snapshot correctly 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-disableGutters MuiContainer-maxWidthLg"
  >
    <div
      class="MuiPaper-root WidgetBox-widgetBox-31 MuiPaper-outlined"
    >
      <div
        class="MuiBox-root MuiBox-root-61"
      >
        <div>
          <form
            class="JSONForm-themedForm-64"
          >
            <div
              class=""
            >
              <label
                class="makeStyles-label-91"
                for="root"
              >
                
              </label>
              <div>
                <div
                  class="unsupported-field"
                >
                  <p>
                    Unsupported field schema
                    <span>
                       for
                       field 
                      <code>
                        root
                      </code>
                    </span>
                    <em>
                      : 
                      Unknown field type undefined
                    </em>
                    .
                  </p>
                  <pre>
                    {
  "payload": [
    {
      "reusableSubProcesses": null,
      "process-id": "Mortgage_Process.MortgageApprovalProcess",
      "associatedEntities": null,
      "container-id": "mortgage-process_1.0.0-SNAPSHOT",
      "package": "com.myspace.mortgage_app",
      "process-name": "MortgageApprovalProcess",
      "dynamic": false,
      "serviceTasks": null,
      "processVariables": null,
      "process-version": "1.0"
    },
    {
      "reusableSubProcesses": null,
      "process-id": "evaluation",
      "associatedEntities": null,
      "container-id": "evaluation_1.0.0-SNAPSHOT",
      "package": "Evaluation.src.main.resources",
      "process-name": "Evaluation",
      "dynamic": false,
      "serviceTasks": null,
      "processVariables": null,
      "process-version": "1"
    },
    {
      "reusableSubProcesses": null,
      "process-id": "itorders.orderhardware",
      "associatedEntities": null,
      "container-id": "itorders_1.0.0-SNAPSHOT",
      "package": "org.jbpm.demo.itorders",
      "process-name": "Order for IT hardware",
      "dynamic": true,
      "serviceTasks": null,
      "processVariables": null,
      "process-version": "1.0"
    },
    {
      "reusableSubProcesses": null,
      "process-id": "Mortgage_Process.MortgageApprovalProcess",
      "associatedEntities": null,
      "container-id": "mortgage-process_1.0.1-SNAPSHOT",
      "package": "com.myspace.mortgage_app",
      "process-name": "MortgageApprovalProcess",
      "dynamic": false,
      "serviceTasks": null,
      "processVariables": null,
      "process-version": "1.0"
    },
    {
      "reusableSubProcesses": null,
      "process-id": "com.redhat.bpms.examples.mortgage.MortgageApplication",
      "associatedEntities": null,
      "container-id": "mortgage_1.0",
      "package": "com.redhat.bpms.examples.mortgage",
      "process-name": "MortgageApplication",
      "dynamic": false,
      "serviceTasks": null,
      "processVariables": null,
      "process-version": "1.0"
    },
    {
      "reusableSubProcesses": null,
      "process-id": "itorders-data.place-order",
      "associatedEntities": null,
      "container-id": "itorders_1.0.0-SNAPSHOT",
      "package": "org.jbpm.demo.itorders",
      "process-name": "place-order",
      "dynamic": false,
      "serviceTasks": null,
      "processVariables": null,
      "process-version": "1.0"
    }
  ],
  "errors": []
}
                  </pre>
                </div>
              </div>
            </div>
            <div
              class="JSONForm-actionButtons-62"
            >
              <button
                class="MuiButtonBase-root-124 MuiButton-root-95 MuiButton-contained-103 MuiButton-containedPrimary-104"
                tabindex="0"
                type="submit"
              >
                <span
                  class="MuiButton-label-96"
                >
                  common.submit
                </span>
                <span
                  class="MuiTouchRipple-root-127"
                />
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
  
</div>
`;

exports[`<ProcessDefinitionContainer /> renders snapshot correctly on error state 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-disableGutters MuiContainer-maxWidthLg"
  >
    <div
      class="MuiPaper-root WidgetBox-widgetBox-167 MuiPaper-outlined"
    >
      <div
        class="MuiBox-root MuiBox-root-197"
      >
        <div>
          <span
            class="MuiSkeleton-root MuiSkeleton-rect JSONFormSkeleton-header-207 MuiSkeleton-pulse"
            style="width: 280px; height: 18px;"
          />
          <hr
            class="MuiDivider-root JSONFormSkeleton-divider-208"
          />
        </div>
        <div
          class="JSONFormSkeleton-item-209"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
            style="width: 80px; height: 15px;"
          />
          <hr
            class="MuiDivider-root JSONFormSkeleton-field-210"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
            style="width: 120px; height: 12px;"
          />
        </div>
        <div>
          <span
            class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
            style="width: 80px; height: 15px;"
          />
          <hr
            class="MuiDivider-root JSONFormSkeleton-field-210"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
            style="width: 160px; height: 12px;"
          />
        </div>
        <div
          class="JSONFormSkeleton-controls-211"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-pulse"
            style="width: 80px; height: 30px;"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiSnackbar-root MuiSnackbar-anchorOriginBottomCenter"
  >
    <div
      class="MuiPaper-root MuiSnackbarContent-root Notification-error-204 MuiPaper-elevation6"
      direction="up"
      role="alert"
      style="opacity: 1; transform: scale(1, 1); transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiSnackbarContent-message"
      >
        <span
          class="Notification-message-201"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root Notification-icon-202 Notification-iconVariant-203"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
            />
          </svg>
          invalid json response body at undefined reason: Unexpected end of JSON input
        </span>
      </div>
      <div
        class="MuiSnackbarContent-action"
      >
        <button
          aria-label="close"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root Notification-icon-202"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
